adding binder custom tree option

choldgraf opened the pull request:
This PR adds the ability to collect jupyter notebooks in a way that preserves their relative paths/folders. It also generates Binder URLs that point to these paths.

## The challenge

To my knowledge, sphinx gallery doesn't have access to the `_downloads` folder until _after_ the site is built. To that extent, I don't think we can modify / duplicate those files after they've been created. I also don't think it's possible to tell the `:download:` directive to put a file in some relative path in `_downloads`. It only dumps it into the root of the directory.

So I think our only option (used in this PR) is kind of hacky: putting the folders/files in `_static` just before the site is built.

So, this PR does:

1. If the binder configuration `static_folder` is defined (relative to `conf.py`) it will trigger the following process:
2. Each ipynb file will be copied to a sub-tree within `<staticfolder>/binder/path/to/ntbk.ipynb` during the build process
3. Binder links will point to this sub-tree rather than the `_downloads` folder

It also adds in the ability to specify `use_lab=True` in the binder config, in which case it will change the end of the link from:

`?filepath=_downloads/myfile.ipynb` to `?urlpath=lab/tree/_downloads/myfile.ipynb`

LMK what folks think! It's a bit hacky but I think our only option. In the docs, we should recommend that people put the output folder under their `gitignore`.

Here's a binder link that shows what the final folder/file structure looks like:

https://mybinder.org/v2/gh/choldgraf/sphinx-gallery/gh-pages?urlpath=lab/tree/_static/binder/auto_examples/plot_colors.ipynb

LMKWYT!

cc @lesteve 

closes #353 #368

choldgraf commented on doc/conf.py:
```diff
@@ -341,10 +341,15 @@ def setup(app):
                                   '../examples/no_output/plot_syntaxerror.py'],
-    'binder': {'org': 'sphinx-gallery',
-               'repo': 'sphinx-gallery.github.io',
+    'binder': {#'org': 'sphinx-gallery',
```
I will update these back to their respective values when we're +1 on a merge...this is just for testing

lesteve commented:
Thanks a lot for the PR! I have not looked at the details yet, but the main question I have is: is there a good reason that we *have* to use `_static`? I would have thought that the variable should be called `notebooks_output_folder` or something similar and that and that the user could chose any folder he wants but I may be missing a sphinx quirk.

lesteve commented:
Also some CI failures seem genuine, but I don't think that is the main problem right now.

choldgraf commented:
@lesteve as far as I could figure out, sphinx won't copy over anything that isn't RST or markdown (e.g., it doesn't copy over anything other than this in the "auto_examples" folder, but correct me if I'm wrong) so the only option is to put it in `_static/myfolder`

If somebody knows a hack around this, lemme know!

lesteve commented:
Can we not copy the notebooks ourselves? This was my naive strategy originally:
* we may be able to know the html output folder (`_build/html` in the sphinx-gallery case): if `app` is a `Sphinx` object, it looks like `app.outdir` is what we want.
* once we know the html output folder we add `notebooks_output_folder` to it so that the notebook are inside the generated html. That means that the notebooks are part of the repo and we can use binder with the appropriate URLs.

lesteve commented:
Something like this, copying the notebooks rather than moving seems safer. Just to be clear, in my mind `notebook_output_dir` is a subfolder or `app.outdir`. Just to take an example in the sphix-gallery case, if I have `notebook_output_dir='my/notebooks'` I should have the notebooks in `_build/html/my/notebooks`.

I agree that this is a bit working around sphinx but I feel this is a bit what sphinx-gallery is already doing to some respect (e.g. the backreferences aka mini-galleries in the reference API doc comes to mind). Of course, if there are ways to do it the sphinx way for this PR or others, this would be more than welcome!

choldgraf commented:
lemme give this a shot and we'll see how it looks :-)

choldgraf commented:
@lesteve what do you think about the latest push? You can now specify an arbitrary folder, and the notebooks will be copied to that folder within `_build`.

e.g.:

http://predictablynoisy.com/sphinx-gallery/auto_examples/plot_seaborn.html#sphx-glr-auto-examples-plot-seaborn-py

I think part of the errors are because I'm using a python 3 function somewhere, I'll try and run that down, but in the meantime WDYT about the current pattern?

Also - do you think that we should actually do this automatically, rather than have people provide a config to enable it? E.g., we could basically make `notebooks_dir` to default to `notebooks`.

Titan-C commented on sphinx_gallery/binder.py:
```diff
@@ -45,20 +45,39 @@ def gen_binder_url(fname, binder_conf):
+            link_base, replace_py_ipynb(relative_link))
+    else:
+        # Assume that we're in the root of _downloads
+        fname = os.path.basename(fname)
```
why do you need this extra `basename`?

choldgraf replied:
I believe that fname here is a full path, so this just grabs the name of the file itself

Titan-C commented on sphinx_gallery/gen_rst.py:
```diff
@@ -709,7 +711,19 @@ def generate_file_rst(fname, target_dir, src_dir, gallery_conf):
+    save_notebook(example_nb, path_ipynb)
+
+    # Copy the ipynb file to the static folder, preserving folder structure
+    if binder_conf.get('notebooks_folder'):
```
I don't like doing the copy part here nor passing the app object this deep, this breaks my desire of making sphinx-gallery more modular. This copy task should happen at the end of the build in the same way we implement `embed_code_links`.
That is have in setup something like this
```python
app.connect('build-finished',copy_jupyter_nb)
```
And then in binder.py the funtion that does the copy.

Titan-C replied:
maybe for the the copy just use an `os.walk` and copy everything that is a notebook.

choldgraf replied:
yeah I think os.walk is the path forward here...lemme give a shot at modularizing this a little bit more.

choldgraf commented:
ok I *think* the latest push should address much of @lesteve 's concerns. It factors out the notebook copying step to the very end as you suggested. Also uses `os.walk` now, so perhaps that means the python 2 builds will be happy again

choldgraf commented:
oops, that was @Titan-C 's concerns not @lesteve 's concerns :-)

Latest version sets `notebooks_folder` to default to `notebooks/`, and lets people change this if they wish.

LMK if there's anything else I should tackle other than making tests pass.

choldgraf commented:
OK I think we're looking good w/ tests and coverage. This is ready for another review, or ready to 🚢

choldgraf commented:
huh - yes I see the same thing too! I didn't check to see whether *extra* files and folders were created haha...lemme check on this. And I agree with `_dir` being consistent with the broader project

choldgraf commented:
OK I think I fixed it up. Here's the my build tree now:

```
doc/_build/html/notebooks/
├── auto_examples
│   ├── no_output
│   │   ├── just_code.ipynb
│   │   ├── plot_raise.ipynb
│   │   ├── plot_strings.ipynb
│   │   └── plot_syntaxerror.ipynb
│   ├── plot_choose_thumbnail.ipynb
│   ├── plot_colors.ipynb
│   ├── plot_exp.ipynb
│   ├── plot_function_identifier.ipynb
│   ├── plot_gallery_version.ipynb
│   ├── plot_quantum.ipynb
│   ├── plot_seaborn.ipynb
│   ├── plot_sys_argv.ipynb
│   ├── plot_unicode_everywhere.ipynb
│   └── sin_func
│       ├── plot_sin_black_background.ipynb
│       └── plot_sin.ipynb
└── tutorials
    ├── plot_notebook.ipynb
    ├── plot_parse.ipynb
    └── seaborn
        └── plot_seaborn_notebook.ipynb
```

also replaced `_folder` with `_dir`

choldgraf commented:
friendly ping :-)

Titan-C commented on doc/advanced_configuration.rst:
```diff
@@ -442,19 +442,27 @@ dictionary following the pattern below::
+         # Optional keys
+         'filepath_prefix': '<prefix>' # A prefix to append to any filepaths in Binder links.
+                            use this if you move your built documentation to a sub-folder of your repository (e.g., "v2.1").
+         'notebooks_dir': '<notebooks-directory-name>' # A directory name where jupyter notebooks that Binder links point to will be moved.
```
I find this hard to read. 
Also, I know this is not  "runnable" code, but I would appreciate that multi-line comments do start with `#`.

Titan-C commented on doc/conf.py:
```diff
@@ -341,10 +341,15 @@ def setup(app):
+               # 'branch': 'master',
+               'dependencies': './binder/requirements.txt',
+               'notebooks_dir': 'notebooks',
+               'use_lab': True
```
Nitpick. Terminate this with a comma.

Titan-C commented on sphinx_gallery/binder.py:
```diff
@@ -95,17 +118,61 @@ def gen_binder_rst(fname, binder_conf):
+    """
+    gallery_dirs = gallery_conf.get('gallery_dirs')
+    if not isinstance(gallery_dirs, (list, tuple)):
+        gallery_dirs = [gallery_dirs]
```
This control is already done in the generation of the gallery. No need to redo it here

Titan-C commented on sphinx_gallery/binder.py:
```diff
@@ -95,17 +118,61 @@ def gen_binder_rst(fname, binder_conf):
+                        app.outdir, notebooks_dir, path_rel)
+                    if not os.path.exists(path_out_dir):
+                        os.makedirs(path_out_dir)
+                    sh.copyfile(os.path.join(i_path, i_file),
```
can we use shutil instead of just sh.

Titan-C commented on sphinx_gallery/binder.py:
```diff
@@ -95,17 +118,61 @@ def gen_binder_rst(fname, binder_conf):
+    _copy_binder_notebooks(app, gallery_conf, binder_conf)
+
+
+def _copy_binder_reqs(app, gallery_conf, binder_conf):
```
Why is there a need to change this function? `gallery_conf` is not used, remove it. Also use `shutil.copy` not `sh.copy`

choldgraf replied:
removed the gallery conf (was from an earlier version of the PR I think). I modified the function so that it has the same general structure as the "copy notebooks" function

Titan-C commented on sphinx_gallery/gen_rst.py:
```diff
@@ -709,7 +710,9 @@ def generate_file_rst(fname, target_dir, src_dir, gallery_conf):
     example_nb = jupyter_notebook(script_blocks)
-    save_notebook(example_nb, replace_py_ipynb(example_file))
+    path_ipynb = replace_py_ipynb(example_file)
+    save_notebook(example_nb, path_ipynb)
```
Why this change? Please keep in mind I also try to get pr #239 and I don't want too many things changing unnecessarily as I need to keep track of the changes.

choldgraf replied:
just cruft from an earlier iteration of this PR, reverting

Titan-C commented on sphinx_gallery/tests/test_gen_gallery.py:
```diff
@@ -273,3 +273,33 @@ def test_collect_gallery_files(config_app, tmpdir):
+project = u'Sphinx-Gallery <Tests>'
+
+sphinx_gallery_conf = {
+    'mod_example_dir' : os.path.join('modules', 'gen'),
```
`mod_example_dir` is deprecated use `backreferences_dir`

Titan-C commented on sphinx_gallery/binder.py:
```diff
@@ -158,3 +227,15 @@ def check_binder_conf(binder_conf):
     return binder_conf
+
+
+def _get_gallery_dir_path(fname, gallery_conf):
```
I don't really understand why this function is needed. Also, I don't really understand what it is trying to do. Can you do a unit test of this and parametrize to many inputs? So that I can get an idea of what this function is trying to do and be sure we need it. I apologize not yet running the code and thus having little idea what goes on in the building of the binder directory.

choldgraf replied:
actually I think that I can accomplish the same thing with `relpath` now that I think about it, lemme give that a shot

Titan-C commented:
This looks good to me. Thank you for the quick refactoring. Happy to see you use `relpath` now and not your own function. It makes the code more readable now. I took the liberty to update the function that copies the jupyter notebooks. Please review it.
This is ready for a merge if everything looks good on @choldgraf live website, then we can restore the old config.

choldgraf commented:
Thanks for the patch! I cleaned up the filtering function a little bit since the multi-part boolean logic was confusing to me. Other than that, I think this is ready to go from my end. Here's a link to the live build:

https://predictablynoisy.com/sphinx-gallery/auto_examples/sin_func/plot_sin.html#sphx-glr-auto-examples-sin-func-plot-sin-py

choldgraf commented:
Sounds good - I am all for beautifully complex list comprehensions, but I must admit it took me like 10 minutes to figure out what that one was doing :-) 

I think the main thing that confused me was these lines:

`entry.endswith('.ipynb') != os.path.isdir(os.path.join(path, entry))`

anyway feel free to suggest alternate namings etc...

Titan-C commented:
> `entry.endswith('.ipynb') != os.path.isdir(os.path.join(path, entry))`

You have to see != as an exclusive OR. Then it works fine and you can
read: either a Jupyter notebook or a directory. Then the NOT in front
transforms it to neither ... nor.

choldgraf commented:
I tried updating the functions name/description/structure to make it a bit more clear...LMK what you think (or feel free to make edits)

lesteve commented:
Sorry I haven't had time to look at this which I think is really important. I'll try to find some time beginning of next week.

choldgraf commented:
friendly ping on this :-)

Titan-C commented:
I'm fine with this one. Just update the config values to point to
sphinx-gallery repo and not yours.

lesteve commented:
I am going to look at this one today.

lesteve commented on sphinx_gallery/binder.py:
```diff
@@ -95,17 +117,76 @@ def gen_binder_rst(fname, binder_conf):
+        gallery_dirs, 'copying binder notebooks...', length=len(gallery_dirs))
+
+    for i_folder in iterator:
+        shutil.copytree(os.path.join(app.srcdir, i_folder),
```
In terms of readability I feel it would be clearer to walk the folder and copy all the .ipynb files explicitly. You would have to change the the beginning of the path from `app.srcdir` to `i_folder` by hand.

In other words if we only want to select `*.ipynb` files, I find the `ignore` API is not that well suited.

Reflecting about this though, I am wondering whether notebooks need other files to be fully functional. One example that comes to mind is if you do an import of a local python module (I seem to remember someone, maybe matplotlib, was doing it).

Titan-C replied:
I first proposed to use os.walk but after looking at @choldgraf's use of it I though it was a very bad idea to push forward, I needed few nested loops and you are joining paths all the time manually. I find the copytree implementation way cleaner. It copies the entire tree structure of the examples but only keeping the notebook files. It may feel counter intuitive to think about the ignore list but it's simpler to extend later on.
The ignore list is easier to extend and test as it is a separate function.

Regarding your use case of needing to copy other files that are not an example, the ignore list makes it easier. Unfortunately we will meet other roadblocks on that area regarding the way Sphinx-Gallery works. Examples are run in the `examples_dirs`, and thus look for file local to that place, in that place we put local modules, data files(like nilearn), or what ever else needed to run the script. But the `gallery_dirs` was only planned for web display, thus has only source code of executed file, rst representation, jupyter notebook representation and output images.

That means if we need to copy other files to give a full reproducible experience where we need local files, we would need to add an extra copy procedure to copy from the `examples_dirs`

Titan-C replied:
No offence is meant to @choldgraf for his implementation. That was probably the way to use os.walk, it just didn't look good to have it.

choldgraf replied:
No worries - I just want a clear and short path forward to get this PR merged. What shall we agree on for this particular point? Keep it the same?

lesteve replied:
I think it is fine to keep it like this for now.

lesteve commented on sphinx_gallery/tests/test_gen_gallery.py:
```diff
@@ -273,3 +273,33 @@ def test_collect_gallery_files(config_app, tmpdir):
+    'gallery_dirs': ['ex'],
+    'binder': {'url': 'http://test1.com', 'org': 'org',
+               'repo': 'repo', 'branch': 'branch',
+               'dependencies': 'requirements.txt'}
```
It would be nice to have a non-default notebook folder here, just to double-check that the notebook output directory is fine.

choldgraf commented on sphinx_gallery/binder.py:
```diff
@@ -95,17 +117,76 @@ def gen_binder_rst(fname, binder_conf):
+        if entry.endswith('.ipynb'):
+            # Don't include ipynb files
+            pass
+        elif (entry != "images") and os.path.isdir(os.path.join(path, entry)):
```
not sure, this one was @Titan-C 's , I was just adapting it w/o changing functionality :-)

Titan-C replied:
We keep images because is an ignore list and I don't care about them. We need the notebooks and directories containing them. The image folder just adds to the noise of folders copied.

choldgraf commented:
I believe that I addressed all of @lesteve 's comments (there are two remaining ones where it wasn't clear if/what I should change). Let me know what you guys think!

lesteve commented:
I ran this on scikit-learn and I am seeing this warning:
```
WARNING: Duplicate file name(s) found. Having duplicate file names will break some links. List of files: ['../examples/tree/plot_iris.py']
```

Unless I am missing something, we are now mirroring the folder structure, so duplicate file names are not a problem anymore and we should remove this logic.

larsoner commented:
IIUC it still creates problems with image file names from one example
affecting the file names for the other (sphinx controls this to avoid
collisions). If so we should keep it.

>

Titan-C commented:
Here we are copying the folder structure to generate the binder links. Be it for single notebook or binder lab. In this way we should not see any problems with links, which was the unpredictable part as name collisions would change the filenames when they were moved to the `_downloads` folder. Regardless of that, keeping this as a warning is very useful and we should keep it.

lesteve commented:
OK fine for leaving it as is. I checked and found https://github.com/sphinx-gallery/sphinx-gallery/pull/244#issuecomment-335531679 for example. Although the warnings was introduced in a binder context but you could argue that it makes sense in general.

larsoner commented:
Yes the problem is more general than the binder context.

>

choldgraf commented:
so.......is there anything for me to do here? :-)

lesteve commented:
I pushed some tweaks, merging, thanks a lot @choldgraf, great stuff !

choldgraf commented:
woooo!!!! 🎉 🎉 🎉 🎉
